// @remix-run/vercel/server.js
/**
 * @remix-run/vercel v1.4.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var node = require('@remix-run/node');

/**
 * A function that returns the value to use as `context` in route `loader` and
 * `action` functions.
 *
 * You can think of this as an escape hatch that allows you to pass
 * environment/platform-specific values through to your loader/action.
 */

/**
 * Returns a request handler for Vercel's Node.js runtime that serves the
 * response using Remix.
 */
function createRequestHandler({
  build,
  getLoadContext,
  mode = process.env.NODE_ENV
}) {
  let handleRequest = node.createRequestHandler(build, mode);
  return async (req, res) => {
    let abortController = new node.AbortController();
    let request = createRemixRequest(req, abortController);
    let loadContext = typeof getLoadContext === "function" ? getLoadContext(req, res) : undefined;
    let response = await handleRequest(request, loadContext);

    if (abortController.signal.aborted) {
      response.headers.set("Connection", "close");
    }

    sendRemixResponse(res, response);
  };
}
function createRemixHeaders(requestHeaders) {
  let headers = new node.Headers();

  for (let key in requestHeaders) {
    let header = requestHeaders[key]; // set-cookie is an array (maybe others)

    if (Array.isArray(header)) {
      for (let value of header) {
        headers.append(key, value);
      }
    } else {
      headers.append(key, header);
    }
  }

  return headers;
}
function createRemixRequest(req, abortController) {
  let host = req.headers["x-forwarded-host"] || req.headers["host"]; // doesn't seem to be available on their req object!

  let protocol = req.headers["x-forwarded-proto"] || "https";
  let url = new URL(req.url, `${protocol}://${host}`);
  let init = {
    method: req.method,
    headers: createRemixHeaders(req.headers),
    abortController,
    signal: abortController === null || abortController === void 0 ? void 0 : abortController.signal
  };

  if (req.method !== "GET" && req.method !== "HEAD") {
    init.body = req;
  }

  return new node.Request(url.href, init);
}
function sendRemixResponse(res, nodeResponse) {
  var _nodeResponse$body;

  let arrays = new Map();

  for (let [key, value] of nodeResponse.headers.entries()) {
    if (arrays.has(key)) {
      let newValue = arrays.get(key).concat(value);
      res.setHeader(key, newValue);
      arrays.set(key, newValue);
    } else {
      res.setHeader(key, value);
      arrays.set(key, [value]);
    }
  }

  res.statusMessage = nodeResponse.statusText;
  res.writeHead(nodeResponse.status, nodeResponse.headers.raw());

  if (Buffer.isBuffer(nodeResponse.body)) {
    res.end(nodeResponse.body);
  } else if ((_nodeResponse$body = nodeResponse.body) !== null && _nodeResponse$body !== void 0 && _nodeResponse$body.pipe) {
    nodeResponse.body.pipe(res);
  } else {
    res.end();
  }
}

exports.createRemixHeaders = createRemixHeaders;
exports.createRemixRequest = createRemixRequest;
exports.createRequestHandler = createRequestHandler;
exports.sendRemixResponse = sendRemixResponse;
