// @remix-run/node/index.js
/**
 * @remix-run/node v1.4.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var sourceMapSupport = require('source-map-support');
var abortController = require('abort-controller');
var fetch = require('./fetch.js');
var formData = require('./formData.js');
var globals = require('./globals.js');
var parseMultipartFormData = require('./parseMultipartFormData.js');
var fileStorage = require('./sessions/fileStorage.js');
var fileUploadHandler = require('./upload/fileUploadHandler.js');
var memoryUploadHandler = require('./upload/memoryUploadHandler.js');
var implementations = require('./implementations.js');
var serverRuntime = require('@remix-run/server-runtime');
var nodeFetch = require('node-fetch');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var sourceMapSupport__default = /*#__PURE__*/_interopDefaultLegacy(sourceMapSupport);

sourceMapSupport__default["default"].install();

Object.defineProperty(exports, 'AbortController', {
  enumerable: true,
  get: function () { return abortController.AbortController; }
});
exports.Request = fetch.Request;
exports.fetch = fetch.fetch;
exports.FormData = formData.FormData;
exports.installGlobals = globals.installGlobals;
exports.unstable_parseMultipartFormData = parseMultipartFormData.parseMultipartFormData;
exports.createFileSessionStorage = fileStorage.createFileSessionStorage;
exports.NodeOnDiskFile = fileUploadHandler.NodeOnDiskFile;
exports.unstable_createFileUploadHandler = fileUploadHandler.createFileUploadHandler;
exports.unstable_createMemoryUploadHandler = memoryUploadHandler.createMemoryUploadHandler;
exports.createCookie = implementations.createCookie;
exports.createCookieSessionStorage = implementations.createCookieSessionStorage;
exports.createMemorySessionStorage = implementations.createMemorySessionStorage;
exports.createSessionStorage = implementations.createSessionStorage;
Object.defineProperty(exports, 'createRequestHandler', {
  enumerable: true,
  get: function () { return serverRuntime.createRequestHandler; }
});
Object.defineProperty(exports, 'createSession', {
  enumerable: true,
  get: function () { return serverRuntime.createSession; }
});
Object.defineProperty(exports, 'isCookie', {
  enumerable: true,
  get: function () { return serverRuntime.isCookie; }
});
Object.defineProperty(exports, 'isSession', {
  enumerable: true,
  get: function () { return serverRuntime.isSession; }
});
Object.defineProperty(exports, 'json', {
  enumerable: true,
  get: function () { return serverRuntime.json; }
});
Object.defineProperty(exports, 'redirect', {
  enumerable: true,
  get: function () { return serverRuntime.redirect; }
});
Object.defineProperty(exports, 'Headers', {
  enumerable: true,
  get: function () { return nodeFetch.Headers; }
});
Object.defineProperty(exports, 'Response', {
  enumerable: true,
  get: function () { return nodeFetch.Response; }
});
